File: cmd/harnest/compile_selection.go
package main

import (
	"bytes"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/pelletier/go-toml/v2"
	"gopkg.in/yaml.v3"
	"harnest.dev/harnest/engine"
)

const compileSelectionFile = "harnest-compile.yaml"

var compileExtraName = regexp.MustCompile(`^[A-Za-z0-9]+([._-][A-Za-z0-9]+)*$`)

type compileSelection struct {
	Extras []string `json:"extras"`
}

// readCompileSelection accepts inert, versioned declarations without loading project packs.
func readCompileSelection(root string) (compileSelection, error) {
	data, err := readRegularDependencyFile(filepath.Join(root, compileSelectionFile))
	if os.IsNotExist(err) {
		return compileSelection{}, nil
	}
	if err != nil {
		return compileSelection{}, err
	}
	decoder := yaml.NewDecoder(bytes.NewReader(data))
	var document map[string]any
	if err = decoder.Decode(&document); err != nil {
		return compileSelection{}, err
	}
	var trailing any
	if err = decoder.Decode(&trailing); err != io.EOF {
		return compileSelection{}, fmt.Errorf("%s must contain one YAML document", compileSelectionFile)
	}
	return parseCompileSelection(document)
}

// parseCompileSelection rejects misspellings rather than silently omitting team requirements.
func parseCompileSelection(document map[string]any) (compileSelection, error) {
	var result compileSelection
	if version, ok := document["version"].(int); !ok || version != 1 {
		return result, fmt.Errorf("%s requires version: 1", compileSelectionFile)
	}
	for key := range document {
		if key != "version" && key != "extras" {
			return result, fmt.Errorf("unknown %s field %q", compileSelectionFile, key)
		}
	}
	var err error
	result.Extras, err = compileSelectionStrings(document, "extras")
	if err != nil {
		return result, err
	}
	return result, validateCompileSelection(result)
}

// compileSelectionStrings distinguishes an omitted list from malformed explicit nulls.
func compileSelectionStrings(document map[string]any, key string) ([]string, error) {
	value, exists := document[key]
	return strictStringList(value, compileSelectionFile+"."+key, !exists)
}

// validateCompileSelection restricts selections to Python optional-dependency group names.
func validateCompileSelection(value compileSelection) error {
	for _, extra := range value.Extras {
		if !compileExtraName.MatchString(extra) {
			return fmt.Errorf("invalid compile extra %q", extra)
		}
	}
	return nil
}

// compileExtraRequirements selects only named root-project extras; base and extension requirements remain mandatory.
func compileExtraRequirements(project string, extras []string) ([]string, error) {
	if len(extras) == 0 {
		return nil, nil
	}
	data, err := readRegularDependencyFile(project)
	if err != nil {
		return nil, err
	}
	var document map[string]any
	if err = toml.Unmarshal(data, &document); err != nil {
		return nil, err
	}
	groups := tomlTable(tomlTable(document["project"])["optional-dependencies"])
	var requirements []string
	for _, extra := range uniqueSorted(extras) {
		values, err := strictStringList(groups[extra], "[project.optional-dependencies]."+extra, false)
		if err != nil {
			return nil, err
		}
		if err := validateCompileRequirements(values); err != nil {
			return nil, err
		}
		requirements = append(requirements, values...)
	}
	return uniqueSorted(requirements), nil
}

// inspectCompileDependencyPlan keeps compile-only selections outside ordinary agent environments.
func inspectCompileDependencyPlan(bundle engine.Bundle) (runtimeDependencyPlan, error) {
	plan, err := inspectRuntimeDependencyPlan(bundle)
	if err != nil {
		return plan, err
	}
	selection, err := readCompileSelection(bundle.Directory)
	if err != nil {
		return plan, err
	}
	plan.CompileRequirements, err = compileExtraRequirements(plan.ProjectFiles[0], selection.Extras)
	return plan, err
}

// inspectProfileDependencyPlan isolates optional compilation dependencies from regular runs.
func inspectProfileDependencyPlan(bundle engine.Bundle, profile environmentProfile) (runtimeDependencyPlan, error) {
	if profile == compileEnvironmentProfile {
		return inspectCompileDependencyPlan(bundle)
	}
	return inspectRuntimeDependencyPlan(bundle)
}

// compileProfile preserves the existing production environment when no optional packages are requested.
func compileProfile(bundle engine.Bundle) (environmentProfile, error) {
	plan, err := inspectCompileDependencyPlan(bundle)
	if err != nil {
		return "", err
	}
	if len(plan.CompileRequirements) > 0 {
		return compileEnvironmentProfile, nil
	}
	return runtimeEnvironmentProfile, nil
}

// validateCompileRequirements prevents PEP 621 entries from becoming requirements-file directives.
func validateCompileRequirements(values []string) error {
	for _, value := range values {
		if strings.ContainsAny(value, "\r\n") || !distributionNamePattern.MatchString(value) {
			return fmt.Errorf("compile extra must contain Python package requirements, got %q", value)
		}
	}
	return nil
}

File: src/harnest/compile_selection.py
"""Inert compile-only selections maintained by agents or team project packs."""
from __future__ import annotations

import json
from pathlib import Path, PurePosixPath
import re
from typing import Any

import yaml

import tomllib

from .server_config import _UniqueKeyLoader

FILENAME = "harnest-compile.yaml"
REPORT = "harnest-build-report.json"
_EXTRA = re.compile(r"[A-Za-z0-9]+([._-][A-Za-z0-9]+)*\Z")


def read_compile_selection(root: Path) -> dict[str, Any]:
    """Validate declarations before authored Python or project pack code can run."""
    path = root / FILENAME
    if not path.exists() and not path.is_symlink():
        return {"version": 1, "extras": []}
    try:
        document = yaml.load(_read_regular_file(path), Loader=_UniqueKeyLoader)
    except yaml.YAMLError as exc:
        raise ValueError(f"invalid {FILENAME}: {exc}") from exc
    _validate_document(document)
    result = {"version": 1, "extras": _string_list(document.get("extras", []), "extras")}
    for extra in result["extras"]:
        if not _EXTRA.fullmatch(extra):
            raise ValueError(f"invalid compile extra: {extra!r}")
    _selected_requirements(root, result["extras"])
    return result


def _validate_document(document: Any) -> None:
    """Reject unsupported versions and unknown keys instead of ignoring team intent."""
    if not isinstance(document, dict):
        raise ValueError(f"{FILENAME} must be a mapping")
    if type(document.get("version")) is not int or document["version"] != 1:
        raise ValueError(f"{FILENAME} requires version: 1")
    if set(document) - {"version", "extras"}:
        raise ValueError(f"unknown {FILENAME} field")


def _string_list(value: Any, label: str) -> list[str]:
    """Canonicalize repeated declarations while retaining strict input types."""
    if not isinstance(value, list) or any(not isinstance(item, str) or not item.strip() for item in value):
        raise ValueError(f"{FILENAME}.{label} must be a list of non-empty strings")
    return sorted(set(value))


def _project_relative_path(value: str) -> Path:
    """Require exact portable relative paths without glob or traversal semantics."""
    path = PurePosixPath(value)
    if not path.parts or path.is_absolute() or str(path) != value or ".." in path.parts:
        raise ValueError(f"dependency file must be an exact project-relative path: {value!r}")
    if any(char in value for char in "\\:\x00*?["):
        raise ValueError(f"invalid dependency file path: {value!r}")
    return Path(value)


def write_compile_report(directory: Path, selection: dict[str, Any] | None) -> None:
    """Record actual artifact bytes and inclusion reasons without inspecting runtime imports."""
    if selection is None:
        return
    files = []
    for path in sorted(directory.rglob("*")):
        if not path.is_file():
            continue
        relative = path.relative_to(directory).as_posix()
        files.append({"path": relative, "size": path.stat().st_size,
                      "reasons": ["authored source" if relative.startswith("source/") else "generated artifact"]})
    report = {"version": 1, "kind": "agent", "selection": selection,
              "logicalBytes": sum(item["size"] for item in files), "files": files,
              "selectedRequirements": _selected_requirements(directory / "source", selection["extras"])}
    (directory / REPORT).write_text(json.dumps(report, indent=2, sort_keys=True) + "\n", encoding="utf-8")


def _selected_requirements(root: Path, extras: list[str]) -> list[str]:
    """Validate direct Python compilation against the same PEP 621 declarations as the CLI."""
    if not extras:
        return []
    dependency_file = "pyproject.toml"
    config = root / "config.yaml"
    if config.is_file():
        settings = yaml.load(config.read_text(encoding="utf-8"), Loader=_UniqueKeyLoader)
        dependency_file = settings.get("spec", {}).get("runtime", {}).get("dependencyFile", dependency_file)
    path = root / _project_relative_path(dependency_file)
    project = tomllib.loads(_read_regular_file(path)).get("project", {})
    groups = project.get("optional-dependencies", {})
    requirements = []
    for extra in extras:
        if extra not in groups:
            raise ValueError(f"unknown compile extra: {extra}")
        requirements.extend(_string_list(groups[extra], f"optional-dependencies.{extra}"))
    return sorted(set(requirements))


def _read_regular_file(path: Path) -> str:
    """Bound configuration reads and reject links before decoding any declarations."""
    if path.is_symlink() or not path.is_file() or path.stat().st_size > 16 * 1024 * 1024:
        raise ValueError(f"{path.name} must be a regular file under 16 MiB")
    return path.read_text(encoding="utf-8")
